From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rezoss <rezoss@unica>
Date: Thu, 16 Jul 2026 12:35:00 +0200
Subject: [PATCH] Force chat SmartReply bootstrap dispatch

The Now Nudge UI can be shown by HoneyBoard while SmartSuggestions only
bootstraps the chat session and never reaches the model-backed SmartReply
request path. Keep the package gate forced for supported chat packages and
dispatch startSmartReply directly from the bootstrap path after seeding the
parsed conversation messages.
---
 .../screenunderstanding/SmartReplyHandler.smali       | 146 +++++++++++++
 1 file changed, 146 insertions(+)

diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
@@ -1878,6 +1878,85 @@
 
     const/4 v4, 0x0
 
+    invoke-static/range {p1 .. p1}, Lcom/samsung/android/smartsuggestions/service/policy/PolicyManager;->getInstance(Landroid/content/Context;)Lcom/samsung/android/smartsuggestions/service/policy/PolicyManager;
+
+    move-result-object v5
+
+    invoke-virtual {v5}, Lcom/samsung/android/smartsuggestions/service/policy/PolicyManager;->getPolicyListSync()Lcom/samsung/android/smartsuggestions/service/policy/Policylist;
+
+    move-result-object v5
+
+    if-eqz v5, :cond_rezoss_start_policy_done
+
+    invoke-virtual {v5}, Lcom/samsung/android/smartsuggestions/service/policy/Policylist;->getSuggestedReplyPackageAllowlist()Ljava/util/List;
+
+    move-result-object v5
+
+    invoke-interface {v5, v2}, Ljava/util/List;->contains(Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-eqz v5, :cond_rezoss_start_policy_done
+
+    goto :cond_rezoss_start_force_chat
+
+    :cond_rezoss_start_policy_done
+    const-string v5, "com.whatsapp"
+
+    invoke-static {v2, v5}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-nez v5, :cond_rezoss_start_force_chat
+
+    const-string v5, "com.whatsapp.w4b"
+
+    invoke-static {v2, v5}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-nez v5, :cond_rezoss_start_force_chat
+
+    const-string v5, "com.samsung.android.messaging"
+
+    invoke-static {v2, v5}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-nez v5, :cond_rezoss_start_force_chat
+
+    const-string v5, "org.telegram.messenger"
+
+    invoke-static {v2, v5}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-nez v5, :cond_rezoss_start_force_chat
+
+    const-string v5, "org.telegram.messenger.web"
+
+    invoke-static {v2, v5}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    if-eqz v5, :cond_rezoss_start_original_enabled
+
+    :cond_rezoss_start_force_chat
+    const/4 v1, 0x1
+
+    sput-boolean v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isNowNudgeOn:Z
+
+    sput-boolean v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
+
+    const-string v5, "SmartReplyHandler[reply]"
+
+    const-string/jumbo v6, "force-enabled chat startSmartReply gate"
+
+    invoke-static {v5, v6}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    goto :goto_0
+
+    :cond_rezoss_start_original_enabled
     if-eqz v1, :cond_0
 
     invoke-virtual {v1}, Ljava/lang/Boolean;->booleanValue()Z
@@ -2416,6 +2495,67 @@
 
     invoke-static {v0, v1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
 
+    sput-object p1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->currentSessionId:Landroid/view/contentcapture/ContentCaptureSessionId;
+
+    invoke-direct {p0, p2, p3, p1, p5}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->handleMessages(Landroid/content/Context;Ljava/util/List;Landroid/view/contentcapture/ContentCaptureSessionId;Landroid/graphics/Rect;)V
+
+    sget-object v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->newMessagesSessionMap:Ljava/util/HashMap;
+
+    invoke-virtual {v1, p1}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v1
+
+    check-cast v1, Ljava/util/List;
+
+    if-eqz v1, :cond_rezoss_bootstrap_seed_done
+
+    move-object v0, v1
+
+    check-cast v0, Ljava/util/Collection;
+
+    invoke-interface {v0}, Ljava/util/Collection;->isEmpty()Z
+
+    move-result v0
+
+    if-eqz v0, :cond_rezoss_bootstrap_seed_done
+
+    move-object v0, p3
+
+    check-cast v0, Ljava/util/Collection;
+
+    invoke-interface {v1, v0}, Ljava/util/List;->addAll(Ljava/util/Collection;)Z
+
+    const-string v0, "SmartReplyHandler[reply]"
+
+    const-string/jumbo v1, "seeded chat SmartReply messages from bootstrap"
+
+    invoke-static {v0, v1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    :cond_rezoss_bootstrap_seed_done
+    sget-object v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->replySentMap:Ljava/util/HashMap;
+
+    invoke-virtual {v1, p1}, Ljava/util/HashMap;->remove(Ljava/lang/Object;)Ljava/lang/Object;
+
+    const/4 v0, 0x1
+
+    sput-boolean v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isNowNudgeOn:Z
+
+    sput-boolean v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
+
+    const-string v0, "SmartReplyHandler[reply]"
+
+    const-string/jumbo v1, "force-dispatched chat SmartReply from bootstrap"
+
+    invoke-static {v0, v1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    invoke-virtual {p4}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-direct {p0, p2, p1, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->startSmartReply(Landroid/content/Context;Landroid/view/contentcapture/ContentCaptureSessionId;Ljava/lang/String;)V
+
+    return-void
+
     :cond_rezoss_bootstrap_whatsapp_done
     new-instance v0, Ljava/lang/StringBuilder;
 
-- 
2.43.0
